File: libraryCoreJava/src/main/java/com/juqitech/moretickets/core/serviceloader/Load.java
package com.juqitech.moretickets.core.serviceloader;

import java.io.IOException;
import java.net.URL;
import java.util.Enumeration;

/**
 * @Author zhiqiang
 * @Email liuzhiqiang@moretickets.com
 * @Date 2019-06-06
 * @Description
 */
class Load {

    private ClassLoader loader;
    private Enumeration<URL> configs = null;

    Load() {
        //初始化加载器
        ClassLoader cl = Thread.currentThread().getContextClassLoader();
        if (null != cl) {
            loader = cl;
        } else {
            loader = ClassLoader.getSystemClassLoader();
        }
    }

    //获取URL
    URL initLoad(String location) throws Exception {
        if (configs == null) {
            try {
                if (loader == null) {
                    configs = ClassLoader.getSystemResources(location);
                } else {
                    configs = loader.getResources(location);
                }
            } catch (IOException x) {
                x.printStackTrace();
            }
        }
        return configs.nextElement();
    }

    ClassLoader getLoader() {
        return loader;
    }
}

File: libraryCore/src/main/java/com/themone/core/serviceloader/Load.kt
package com.themone.core.serviceloader

import java.io.IOException
import java.net.URL
import java.util.Enumeration

/**
 * @Author zhiqiang
 * @Email liuzhiqiang@theone.com
 * @Date 2019-06-06
 * @Description
 */
internal class Load {

    var loader: ClassLoader? = null
        private set
    private var configs: Enumeration<URL>? = null

    init {
        //初始化加载器
        val cl = Thread.currentThread().contextClassLoader
        if (null != cl) {
            loader = cl
        } else {
            loader = ClassLoader.getSystemClassLoader()
        }
    }

    //获取URL
    @Throws(Exception::class)
    fun initLoad(location: String): URL {
        if (configs == null) {
            try {
                if (loader == null) {
                    configs = ClassLoader.getSystemResources(location)
                } else {
                    configs = loader!!.getResources(location)
                }
            } catch (x: IOException) {
                x.printStackTrace()
            }

        }
        return configs!!.nextElement()
    }
}
